pilot/cmd/pilot-agent/readiness.go

```diff
@@ -0,0 +1,153 @@
+// Copyright 2018 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"net/http"
+
+	"errors"
+
+	"time"
+
+	adminapi "github.com/envoyproxy/go-control-plane/envoy/admin/v2alpha"
+	"github.com/gogo/protobuf/jsonpb"
+	proto "github.com/gogo/protobuf/types"
+	"github.com/hashicorp/go-multierror"
+)
+
+const (
+	clustersMapKey   = "clusters"
+	listenersMapKey  = "listeners"
+	readinessTimeout = time.Second * 5
+)
+
+var (
+	checkFuncs = []checkFunc{
+		checkHasDynamicActiveClusters,
+		checkHasDynamicActiveListeners,
+	}
+)
+
+type checkFunc func(configDump *adminapi.ConfigDump) error
+
+// checkReady polls the Envoy admin port to determine if the Envoy instance has received its configuration from Pilot.
+func checkReady(adminPort int) (err error) {
+	// Add a prefix to any returned errors.
+	defer func() {
+		if err != nil {
+			err = multierror.Prefix(err, "Envoy proxy is NOT ready.")
+		}
+	}()
+
+	var configDump *adminapi.ConfigDump
+	configDump, err = getConfigDump(adminPort)
```

Comment:
@costinm you mentioned not using config_dump ... what do you propose as an alternative?

Reply:
I don't believe that using Envoy's health check is sufficient. What we need is to determine when Envoy has received at least one ADS update from pilot ... or more specifically, that dynamic clusters and listeners have been received.

Reply:
You can use the stats then, no need to dump entire config.

Reply:
+1.. dont check for config. dump stats and check for atleast one successful lds,cds update

Reply:
Ready isn't just for the initial boot of the pod, it's on going - so I don't know if "at least one successful...." cuts it, as that wouldn't cover stale state?

Reply:
Stale state happens when proxy is disconnected from Pilot. At this point proxy will continue to use stale data.

There is no timestamp in `/stats`, however
@nmittler we can use `/stats/cluster.xds-grpc.upstream_cx_rx_bytes_total` to see how much data is xferred, and we can keep track of last change in the agent.
If envoy actually provided timestamp of last xds update, it would be great.

Reply:
I've taken a look at the available stats, and I suspect we're most interested in `xds-grpc` cluster.  After chatting with @mandarjog, I think what we want to do is to track whether or not the connection to Pilot is `active` and if so, whether we've received an update since it's become active.

@PiotrSikora would you be able to help us understand the meaning (and range) of the following stats?

```
cluster.xds-grpc.internal.upstream_rq_200: 1
cluster.xds-grpc.internal.upstream_rq_2xx: 1
cluster.xds-grpc.internal.upstream_rq_503: 1
cluster.xds-grpc.internal.upstream_rq_5xx: 1
cluster.xds-grpc.lb_healthy_panic: 1
cluster.xds-grpc.max_host_weight: 1
cluster.xds-grpc.membership_change: 1
cluster.xds-grpc.membership_healthy: 1
cluster.xds-grpc.membership_total: 1
cluster.xds-grpc.ssl.ciphers.ECDHE-RSA-AES128-GCM-SHA256: 1
cluster.xds-grpc.ssl.handshake: 1
cluster.xds-grpc.update_attempt: 27
cluster.xds-grpc.update_success: 27
cluster.xds-grpc.upstream_cx_active: 1
cluster.xds-grpc.upstream_cx_http2_total: 1
cluster.xds-grpc.upstream_cx_none_healthy: 1
cluster.xds-grpc.upstream_cx_rx_bytes_buffered: 66624
cluster.xds-grpc.upstream_cx_rx_bytes_total: 763842
cluster.xds-grpc.upstream_cx_total: 1
cluster.xds-grpc.upstream_cx_tx_bytes_buffered: 0
cluster.xds-grpc.upstream_cx_tx_bytes_total: 579732
cluster.xds-grpc.upstream_rq_200: 1
cluster.xds-grpc.upstream_rq_2xx: 1
cluster.xds-grpc.upstream_rq_503: 1
cluster.xds-grpc.upstream_rq_5xx: 1
cluster.xds-grpc.upstream_rq_active: 1
cluster.xds-grpc.upstream_rq_total: 1
```

Reply:
For the cluster stats, you can find the description [here](https://www.envoyproxy.io/docs/envoy/v1.7.0/configuration/cluster_manager/cluster_stats), but I don't believe you should to use that.

You probably want to look at [`listener_manager`](https://www.envoyproxy.io/docs/envoy/v1.7.0/configuration/listeners/stats#listener-manager) and [`cluster_manager`](https://www.envoyproxy.io/docs/envoy/v1.7.0/configuration/cluster_manager/cluster_stats) stats instead, i.e.
```
$ curl localhost:9901/stats | grep _manager
cluster_manager.active_clusters: 1000
cluster_manager.cluster_added: 1000
cluster_manager.cluster_modified: 0
cluster_manager.cluster_removed: 0
cluster_manager.cluster_updated: 0
cluster_manager.cluster_updated_via_merge: 0
cluster_manager.update_merge_cancelled: 0
cluster_manager.update_out_of_merge_window: 0
cluster_manager.warming_clusters: 0
listener_manager.listener_added: 1000
listener_manager.listener_create_failure: 0
listener_manager.listener_create_success: 1000
listener_manager.listener_modified: 0
listener_manager.listener_removed: 0
listener_manager.total_listeners_active: 1000
listener_manager.total_listeners_draining: 0
listener_manager.total_listeners_warming: 0
```

Reply:
Thanks, @PiotrSikora ... that's helpful.  Given all of that, I think the logic for readiness should be something like this:

```golang
type snapshot struct {
  active bool // cluster.xds-grpc_upstream_cx_active >= 1
  t time.Time // may need this to switch to !ready after going inactive for a time
  
  cdsVersion string // cluster_manager.cds.version
  cdsUpdates int // cluster_manager.cds.update_success
  warmingClusters int // cluster_manager.warming_clusters

  ldsVersion string // listener_manager.lds.version
  ldsUpdates int // listener_manager.lds.update_success
  warmingListeners int // listener_manager.total_listeners_warming
}

func (s *snapshot) isReady() bool {
  // Purposely leaving active state out of this check.
  return s.cdsUpdates > 0 && s.warmingClusters == 0 && s.ldsUpdates > 0 && s.warmingListeners == 0
}

func (s *snapshot) isUpdatedSince(old *snapshot) bool {
  return s.cdsVersion != old.cdsVersion && s.ldsVersion != old.ldsVersion
}

func newSnapshot(...) *snapshot {
  ...
}

type probe struct {
  ...
  lastReady *snapshot
  ready bool
}

func (p *probe) isReady() bool {
  s := newSnapshot(...)

  // Case 1: Wait for the first ready snapshot
  if p.lastReady == nil {
    if s.active && s.isReady() {
      // Only save the first ready snapshot
      p.lastReady = s
      p.ready = true
    }
    return p.ready
  }

  // Case 2: the last result was not ready
  if !p.ready {
    if s.active {
      if s.isReady() && s.isUpdatedSince(p.lastReady) {
        // Switch back to ready.
        p.lastReady = s
        p.ready = true
        return true
      }
      // Discard this snapshot even though it's active, since
      // it hasn't been updated since going active.
    }

    // Continue publishing ready up to a time threshold
    p.ready = s.t < p.lastReady.Add(staleThreshold)
    return p.ready
  }

  // Case 3: last snapshot was active
  if s.active {
    p.lastReady = s
    return true
  } else {
    p.ready = s.t < p.lastReady.Add(staleThreshold)
    return p.ready
  }
}
```